### PR TITLE
:construction: Implemented plugin based author model in API v2

### PR DESCRIPTION
**Pull request by naz**
refs #10124

- Author model returns only users that have published posts

This is just a first go with Authors API v2 endpoint only. More to come tomorrow once this approach is discussed :)

### CODE REVIEW

File: core/server/models/plugins/has-posts.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const debug = _debug('ghost-query');
+
+const addHasPostsWhere = function () {
+    const tableName = _.result(this, 'tableName');
+    const comparisonField = `${tableName}.id`;
+
+    this.query(function (qb) {
+        return qb.whereIn(comparisonField, function () {
+            const innerQb = this
+                .distinct('posts_authors.author_id')
```
Comment (kirrg001) [inline core/server/models/plugins/has-posts.js]: I assume this logic will become dynamic to handle authors and tags.

File: core/server/models/plugins/has-posts.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                .select()
+                .from('posts_authors')
+                .whereRaw(`posts_authors.author_id = ${comparisonField}`)
+                .join('posts', 'posts.id', 'posts_authors.post_id')
+                .andWhere('posts.page', '=', false)
```
Comment (kirrg001) [inline core/server/models/plugins/has-posts.js]: IMO authors can write posts or pages.

**Comment by kirrg001** [comment]
Code looks good in general 👍 

Just leaving some thoughts:
- i think we have to clarify the sitemap service before shipping any of this change? if we don't return authors without posts, then these authors will 404 for v2, but sitemap service will still show the urls
- group model representations of existing models in a sub folder?

File: core/server/models/plugins/has-posts.js (hunk 2 above)
Comment (gargol) [reply core/server/models/plugins/has-posts.js]: Hm... I was basing thing on the filter applied [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/plugins/include-count.js#L22-L23). Not sure which way it should be done. But It makes sense that any kind of publication should count. Should we include `unpublished` as well?

File: core/server/models/plugins/has-posts.js (hunk 1 above)
Comment (gargol) [reply core/server/models/plugins/has-posts.js]: Yeap, similar to how relations are configured :)

File: core/server/models/plugins/has-posts.js (hunk 2 above)
Comment (kirrg001) [reply core/server/models/plugins/has-posts.js]: no only published 👍

**Comment by gargol** [comment]
> group model representations of existing models in a subfolder?
This is a good question. For now, there is just one model, so maybe too early of an optimization, but definitely would go with something to separate models that just extend existing ones.

**Comment by gargol** [comment]
Notes from a discussion we had with @kirrg001 about further approach:
1. [ ] Call 'hasPosts' plugin straight from knex during UrlService initialization
2. [ ] Introduce 2nd config for resources - we will maintain 2 configs apiV01 and apiV2 as there's no better approach to separate logic at the moment
3. [ ] Double check if cold boot happens when theme provided ghost-api value changes  (logic similar to `releaseResourcesOnly`)
4. [ ]  Do a UrlService data reset only when the version is changed
5. [ ] Add a comment in the codebase about long-term plan behind resource

**Review by kirrg001 — CHANGES_REQUESTED** [review]
This approach does not work with the get helper.

e.g. {{#get "tags"}}

It will query the tags controller and not the tags-public controller.

**Comment by kirrg001** [comment]
> It will query the tags controller and not the tags-public controller.

You probably just need to tweak the get helper.
If we solve #10106, it will be easier to handle this.

**Comment by kirrg001** [comment]
@gargol Could you please rebase against master? Thanks 😃

File: core/server/helpers/get.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ const RESOURCES = {
         resource: 'posts'
     },
     tags: {
-        alias: 'tags',
+        alias: 'tagsPublic',
```
Comment (kirrg001) [inline core/server/helpers/get.js]: yey so simple 😃

File: core/server/services/themes/index.js [hunk 4]
```diff
@@ -83,8 +83,22 @@ module.exports = {
         // no need to check the score, activation should be used in combination with validate.check
         // Use the two theme objects to set the current active theme
         try {
+            let previousGhostAPI;
+
+            if (this.getActive()) {
```
Comment (kirrg001) [inline core/server/services/themes/index.js]: We ask if we have an active theme, but if not, we just call `this.getActive()` 🤔 
See line 93.

Comment (gargol) [reply core/server/services/themes/index.js]: Yes, but there's:
```
active.set(loadedTheme, checkedTheme, error);
```
between these two calls, which changes the context a bit. `previousGhostAPI` would be `undefined` when starting up the instance first time ;)

Comment (kirrg001) [reply core/server/services/themes/index.js]: Ah yeah true 👍

File: core/server/services/url/UrlService.js [hunk 5]
```diff
@@ -215,17 +223,23 @@ class UrlService {
             .getValue(options);
     }
 
-    reset() {
+    init() {
+        this.resources.fetchResources();
+    }
+
+    reset(removeListeners = true) {
```
Comment (kirrg001) [inline core/server/services/url/UrlService.js]: Let's use the options pattern? We use it everywhere. This avoid having functions with x arguments and you need to pass null or default values.

> reset(options = {removeListeners: true}) {

Comment (gargol) [reply core/server/services/url/UrlService.js]: Sure :sunglasses:

File: core/server/services/url/UrlService.js [hunk 6]
```diff
@@ -65,6 +68,11 @@ class UrlService {
         this.urlGenerators.push(urlGenerator);
     }
 
+    _onThemeChangedListener() {
+        this.reset(true);
```
Comment (kirrg001) [inline core/server/services/url/UrlService.js]: I think this must be false. That's you still run into trouble when switching versions?
Just a guess :)

**Comment by kirrg001** [comment]
Found two more bugs:

### 1

- v2 is enabled
- i have 0 tags listed on the sitemap, because none of my posts have tags
- now i add a tag to a post
- sitemaps are not being refreshed
- same appears for when i remove a tag from a post and the tag is not assigned to any post

I assume this not trivial to solve, because we will receive the url event for the post, but we don't have a trigger that the tag url is now available.

### 2

If i have a tag which is not assigned to any post, i won't receive the tag via the content API, but i can serve the tag on the site.

To reproduce:

- create a tag
- assign this tag to no posts (tag has no posts)
- http://localhost:2368/ghost/api/v2/content/tags/?key=lol won't expose this tag
- serve /tags/xxxx

If you do the same with an author. It works for both cases.

File: core/server/services/url/UrlService.js [hunk 7]
```diff
@@ -215,17 +223,23 @@ class UrlService {
...
         this.queue.reset();
         this.resources.reset();
 
-        this._onQueueStartedListener && this.queue.removeListener('started', this._onQueueStartedListener);
-        this._onQueueEndedListener && this.queue.removeListener('ended', this._onQueueEndedListener);
-        this._onRouterAddedListener && common.events.removeListener('router.created', this._onRouterAddedListener);
+        if (!options.keepListeners) {
+            this._onQueueStartedListener && this.queue.removeListener('started', this._onQueueStartedListener);
+            this._onQueueEndedListener && this.queue.removeListener('ended', this._onQueueEndedListener);
+            this._onRouterAddedListener && common.events.removeListener('router.created', this._onRouterAddedListener);
```
Comment (kirrg001) [inline core/server/services/url/UrlService.js]: You probably need to deregister the new event too 

> this._onThemeChangedListener

File: core/server/services/url/UrlService.js (hunk 6 above)
Comment (kirrg001) [reply core/server/services/url/UrlService.js]: 🤣

**Pull request merged**